src/libcryptobox/cryptobox.c

```diff
@@ -706,18 +755,18 @@ bool rspamd_cryptobox_verify(const unsigned char *sig,
 
 
 #if OPENSSL_VERSION_MAJOR >= 3
-		EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(CRYPTOBOX_CURVE_NID, ENGINE_F_ENGINE_NEW);
-		g_assert(pctx != NULL);
+		OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
+		EVP_PKEY *pkey = EVP_PKEY_new_raw_public_key_ex(libctx, "ED25519", NULL, pk, sizeof(rspamd_sk_t));
```

Comment:
Why `ed25519`? For this curve we use libsodium functions. Openssl is responsible for p256 operations only.

Reply:
The `new_raw_private_key` function works only with `Ed25519`, `Ed448`, `X25519` and `X448`. Maybe I don't understand something vital about PKEY, but there is no way to create PKEY out of buffer `pk` or `sk` of type EC.

Reply:
I don't get it: so if this function can't create p256 keys we need to change p256 to some other curve? No it's not acceptable, we need to create elliptic keys exactly as we did with older OpenSSL, we cannot simply change the curve. Also, I don't think that `EVP_PKEY_CTX_new` is now deprecated - why can't we leave it as is?

Reply:
In fact, we just need to replace `EC_KEY_new_by_curve_name` with `EVP_EC_gen`

Reply:
https://www.openssl.org/docs/manmaster/man7/EVP_PKEY-EC.html